daemon/k8s/pods.go

```diff
@@ -0,0 +1,135 @@
+	nodeTypes "github.com/cilium/cilium/pkg/node/types"
+	"github.com/cilium/cilium/pkg/time"
+)
+
+// LocalPod is Cilium's internal model of the pods running on this node.
+type LocalPod struct {
+	*slim_corev1.Pod
+
+	// UpdatedAt is the time when [LocalPod] was last updated, e.g. it
+	// shows when the pod change was received from the api-server.
+	UpdatedAt time.Time `json:"updatedAt" yaml:"updatedAt"`
+}
+
+func (p LocalPod) TableHeader() []string {
+	return []string{
+		"Name",
+		"HostNetwork",
+		"PodIPs",
+		"Containers",
+		"Phase",
+		"Age",
+	}
+}
+
+func (p LocalPod) TableRow() []string {
+	podIPs := make([]string, len(p.Status.PodIPs))
+	for i := range p.Status.PodIPs {
+		podIPs[i] = p.Status.PodIPs[i].IP
+	}
+	containers := make([]string, len(p.Spec.Containers))
+	for i, cont := range p.Spec.Containers {
+		ports := make([]string, len(cont.Ports))
+		for i, port := range cont.Ports {
+			if port.Name != "" {
+				ports[i] = fmt.Sprintf("%d/%s (%s)", port.ContainerPort, string(port.Protocol), port.Name)
+			} else {
+				ports[i] = fmt.Sprintf("%d/%s", port.ContainerPort, string(port.Protocol))
+			}
+		}
+		contName := cont.Name
+		if len(ports) > 0 {
+			contName += " (" + strings.Join(ports, ",") + ")"
+		}
+		containers[i] = contName
+	}
+	return []string{
+		p.Namespace + "/" + p.Name,
+		strconv.FormatBool(p.Spec.HostNetwork),
+		strings.Join(podIPs, ", "),
+		strings.Join(containers, ", "),
+		string(p.Status.Phase),
+		duration.HumanDuration(time.Since(p.UpdatedAt)),
+	}
+}
+
+const (
+	PodTableName = "k8s-pods"
+)
+
+var (
+	PodNameIndex = newNameIndex[LocalPod]()
+
+	PodTableCell = cell.Group(
+		cell.ProvidePrivate(NewPodTable),
+		cell.Provide(statedb.RWTable[LocalPod].ToTable),
+		cell.Invoke(registerPodReflector),
+	)
+)
+
+func PodByName(namespace, name string) statedb.Query[LocalPod] {
+	return PodNameIndex.Query(namespace + "/" + name)
+}
+
+func NewPodTable(db *statedb.DB) (statedb.RWTable[LocalPod], error) {
+	tbl, err := statedb.NewTable(
+		PodTableName,
+		PodNameIndex,
+	)
+	if err != nil {
+		return nil, err
+	}
+	return tbl, db.RegisterTable(tbl)
+}
+
+func registerPodReflector(jg job.Group, db *statedb.DB, cs client.Clientset, pods statedb.RWTable[LocalPod]) error {
+	if !cs.IsEnabled() {
+		return nil
+	}
+	cfg := podReflectorConfig(cs, pods)
+	return k8s.RegisterReflector(jg, db, cfg)
+}
+
+func podReflectorConfig(cs client.Clientset, pods statedb.RWTable[LocalPod]) k8s.ReflectorConfig[LocalPod] {
+	lw := utils.ListerWatcherWithModifiers(
+		utils.ListerWatcherFromTyped(cs.Slim().CoreV1().Pods("")),
+		func(opts *metav1.ListOptions) {
+			opts.FieldSelector = fields.ParseSelectorOrDie("spec.nodeName=" + nodeTypes.GetName()).String()
+		})
+	return k8s.ReflectorConfig[LocalPod]{
+		Name:          reflectorName,
+		Table:         pods,
+		ListerWatcher: lw,
+		Transform: func(_ statedb.ReadTxn, obj any) (LocalPod, bool) {
+			pod, ok := obj.(*slim_corev1.Pod)
+			if !ok {
+				return LocalPod{}, false
```

Comment:
Out of curiosity, but this should never happen? I.e., the transformer will receive only Pod objs?

Reply:
Yeah shouldn't happen.